### PR TITLE
:tada: Add ability to change shadows' order and place new shadows at t…

### PR DESCRIPTION
**Pull request by akshay-gupta7**
:tada: Add ability to change shadows' order and place new shadows at the top by default #2512 

* With this feature, user can now drag and change the order of shadows just like they did for strokes
* Also, the new shadow added will be at the top by default

### CODE REVIEW

File: frontend/src/app/main/ui/workspace/sidebar/options/menus/shadow.cljs [hunk 1]
```diff
@@ -64,6 +66,21 @@
         select-text
         (fn [ref] (fn [_] (dom/select-text! (mf/ref-val ref))))
 
+        on-drop
+        (fn [_ data]
+          (on-reorder (:index data)))
+
+        [dprops dref] (if (some? on-reorder)
```
Comment (niwinz) [inline frontend/src/app/main/ui/workspace/sidebar/options/menus/shadow.cljs:73]: You really can't use hooks on this manner, the hooks can't be on a conditional. They should be always defined as static definitions and always in the same order. I just comment it for you information, I will change it my self.

**Comment by niwinz** [comment]
The same here. I have created a new PR with your commit and some improvements.

**Comment by akshay-gupta7** [comment]
> The same here. I have created a new PR with your commit and some improvements.

Thank you @niwinz for reviewing. 
Also, I saw your comment on the design limitations of the numeric input component. I want to work on improving the numeric component. I would like your high level inputs on how to go about achieving the same.

Regards

File: frontend/src/app/main/ui/workspace/sidebar/options/menus/shadow.cljs (hunk 1 above)
Comment (akshay-gupta7) [reply frontend/src/app/main/ui/workspace/sidebar/options/menus/shadow.cljs:73]: Thank you for highlighting this. Will keep this in mind going forward

**Comment by niwinz** [comment]
The main idea on creating a wrapper component to the native element is that should behave very similar to the native
Right now, the event callbacks are called with value, but with no event, so I can't access the dom node. If I need dom need to use refs, that is ok for some cases but other they are completely unnecessary. 

Also, there are no way avoid re-render the component, because of the callbacks handling limitation (you need create a new function on each render).

I guess this explanation is not very clear, because the issues are not evident at simple look.
In any case this is just FYI, i will go to address it myself once I finish my current ongoing task.

**Comment by akshay-gupta7** [comment]
> The main idea on creating a wrapper component to the native element is that should behave very similar to the native Right now, the event callbacks are called with value, but with no event, so I can't access the dom node. If I need dom need to use refs, that is ok for some cases but other they are completely unnecessary.
> 
> Also, there are no way avoid re-render the component, because of the callbacks handling limitation (you need create a new function on each render).
> 
> I guess this explanation is not very clear, because the issues are not evident at simple look. In any case this is just FYI, i will go to address it myself once I finish my current ongoing task.

Yes I get this to some extent. I faced a similar issue while trying to apply on-focus: select-all on these numeric-input components(#3179), so as a workaround I put these events on the wrapper of these elements, which is fine as a workaround, but going forward we will need to address this.
Thank you for taking the time to throw some light on this.
Regards

**Pull request closed without merging**